src/target/ext/host_io.rs

```diff
@@ -0,0 +1,228 @@
+        None
+    }
+    /// Enable readlink operation.
+    #[inline(always)]
+    fn enable_readlink(&mut self) -> Option<HostIoReadlinkOps<Self>> {
+        None
+    }
+    /// Enable setfs operation.
+    #[inline(always)]
+    fn enable_setfs(&mut self) -> Option<HostIoSetfsOps<Self>> {
+        None
+    }
+}
+
+define_ext!(HostIoOps, HostIo);
+
+/// Nested Target Extension - Host I/O open operation.
+pub trait HostIoOpen: HostIo {
+    /// Close the open file corresponding to fd and return 0, or -1 if an error
+    /// occurs.
+    fn open(
+        &mut self,
+        filename: &[u8],
+        flags: HostOpenFlags,
+        mode: HostMode,
+    ) -> TargetResult<i32, Self>;
+}
+
+define_ext!(HostIoOpenOps, HostIoOpen);
+
+/// Nested Target Extension - Host I/O close operation.
+pub trait HostIoClose: HostIo {
+    /// Close the open file corresponding to fd and return 0, or -1 if an error
+    /// occurs.
+    fn close(&mut self, fd: i32) -> TargetResult<i64, Self>;
+}
+
+define_ext!(HostIoCloseOps, HostIoClose);
+
+/// Nested Target Extension - Host I/O pread operation.
+pub trait HostIoPread: HostIo {
+    /// Read data from the open file corresponding to fd. Up to count bytes will
+    /// be read from the file, starting at offset relative to the start of the
+    /// file. The target may read fewer bytes; common reasons include packet
+    /// size limits and an end-of-file condition. The number of bytes read is
+    /// returned. Zero should only be returned for a successful read at the end
+    /// of the file, or if count was zero.
+    fn pread(
+        &mut self,
+        fd: i32,
+        count: <Self::Arch as Arch>::Usize,
+        offset: <Self::Arch as Arch>::Usize,
+        output: HostIoOutput<'_>,
+    ) -> TargetResult<(), Self>;
+}
+
+define_ext!(HostIoPreadOps, HostIoPread);
+
+/// Nested Target Extension - Host I/O pwrite operation.
+pub trait HostIoPwrite: HostIo {
+    /// Write data (a binary buffer) to the open file corresponding to fd. Start
+    /// the write at offset from the start of the file. Unlike many write system
+    /// calls, there is no separate count argument; the length of data in the
+    /// packet is used. ‘vFile:pwrite’ returns the number of bytes written,
+    /// which may be shorter than the length of data, or -1 if an error
+    /// occurred.
+    fn pwrite(
+        &mut self,
+        fd: i32,
+        offset: <Self::Arch as Arch>::Usize,
+        data: &[u8],
+    ) -> TargetResult<i32, Self>;
+}
+
+define_ext!(HostIoPwriteOps, HostIoPwrite);
+
+/// Nested Target Extension - Host I/O fstat operation.
+pub trait HostIoFstat: HostIo {
+    /// Get information about the open file corresponding to fd. On success the
+    /// information is returned as a binary attachment and the return value is
+    /// the size of this attachment in bytes. If an error occurs the return
+    /// value is -1.
+    fn fstat(&mut self, fd: i32, output: HostIoOutput<'_>) -> TargetResult<i32, Self>;
```

Comment:
Unlike `pread`, where the output can be an arbitrarily large unstructured hunk of data, we know _exactly_ the size and structure of the output of this method. Instead of using a callback here, we should just have the method return a `HostIoResult<HostIoStat>`, where `HostIoStat` is a Rust struct that matches https://sourceware.org/gdb/current/onlinedocs/gdb/struct-stat.html#struct-stat.

`gdbstub` will then serialize `HostIoStat` on behalf of the implementer.